### PR TITLE
[Enchance] Add filter rules of Mosaic transform

### PR DESCRIPTION
**Pull request by zhaojinjian0000**
## Motivation

Fix two errors of Masaic.

## Modification

1. self.img_scale denotes (h,w). The "self.img_scale[1]"  should be changed to "self.img_scale[0]" in the formula of "y2" at line 2134 and 2135 when " loc == 'bottom_left' ". 
```python
                             min(self.img_scale[0] * 2, center_position_xy[1] +
                                 img_shape_wh[1])
```
2. To avoid NaN loss during training. The code should get rid of the mosaic_bboxes with small area (even zero area). In this submission, after line 2081, the bboxes with a short side smaller than 3 are abandoned.
```python
            mosaic_filter = np.prod(mosaic_bboxes[:, 2:4] - \
                                    mosaic_bboxes[:, 0:2] > 2,  axis=1) == 1
            mosaic_bboxes = mosaic_bboxes[mosaic_filter]
            mosaic_labels = mosaic_labels[mosaic_filter]
```

### CODE REVIEW

**Comment by Yangr116** [comment]
I met the loss nan error in the training (cascade r2-101), and I modified the code according to the second modification, then the loss was normal.

File: mmdet/datasets/pipelines/transforms.py [hunk 1]
```diff
@@ -2131,7 +2136,7 @@ def _mosaic_combine(self, loc, center_position_xy, img_shape_wh):
             x1, y1, x2, y2 = max(center_position_xy[0] - img_shape_wh[0], 0), \
                              center_position_xy[1], \
                              center_position_xy[0], \
-                             min(self.img_scale[1] * 2, center_position_xy[1] +
+                             min(self.img_scale[0] * 2, center_position_xy[1] +
```
Comment (hhaAndroid) [inline mmdet/datasets/pipelines/transforms.py:2139]: Thank you!

File: mmdet/datasets/pipelines/transforms.py [hunk 2]
```diff
@@ -2079,6 +2079,11 @@ def _mosaic_transform(self, results):
             mosaic_bboxes[:, 1::2] = np.clip(mosaic_bboxes[:, 1::2], 0,
                                              2 * self.img_scale[0])
             mosaic_labels = np.concatenate(mosaic_labels, 0)
+            
+            mosaic_filter = np.prod(mosaic_bboxes[:, 2:4] - \
```
Comment (hhaAndroid) [inline mmdet/datasets/pipelines/transforms.py:2083]: If you want to add filtering rules, it is recommended to create a new internal method, and then put the hyperparameters in the initialization method.

```python
def __init__(self,
                 img_scale=(640, 640),
                 center_ratio_range=(0.5, 1.5),
                 min_bbox_size,
                 pad_val=114):
        self.min_bbox_size=min_bbox_size


    def _filter_box_candidates(self, bbox):

        ....
        return ((w > self.min_bbox_size)
                & (h > self.min_bbox_size)
              
```

Comment (zhaojinjian0000) [reply mmdet/datasets/pipelines/transforms.py]: Thank you for your kindly reminding and advice. I have created a new internal method "_filter_box_candidates" and put the hyperparameters "min_bbox_size" in the initialization method. 
(The function Shear._filter_invalid at <mmdet/datasets/pipelines/auto_augment.py> are reference to the created internal method "_filter_box_candidates".)

File: mmdet/datasets/pipelines/transforms.py [hunk 3]
```diff
@@ -1964,16 +1964,20 @@ class Mosaic:
...
            output. Default to (0.5, 1.5).
+        min_bbox_size (int | float): The minimum pixel for filtering
+            invalid bboxes after the mosaic pipeline. Default to 1.
         pad_val (int): Pad value. Default to 114.
     """
 
     def __init__(self,
                  img_scale=(640, 640),
                  center_ratio_range=(0.5, 1.5),
+                 min_bbox_size=1,
```
Comment (hhaAndroid) [inline mmdet/datasets/pipelines/transforms.py:1975]: In order not to affect the original logic, the default value should be set to 0. And only execute when it is greater than 0.

Comment (zhaojinjian0000) [reply mmdet/datasets/pipelines/transforms.py:1975]: Thank you for your kindly reminding and advice.

The added filtering rule aims to fix the BUG of NAN loss during traning. 
## The original code
In Mosaic, the sub image are cropped if the sub image is larger than mosaic patch. However, some bboxes are translated out of image completely (such as bbox2 in figure 1) and some bboxes are translated out of image partly  (such as bbox1 in figure 1).
 ```python
             figure 1. before clip 
            +----------------------+
            |                      |
            |                      |    
            |                      |   +----+
            |                      |   |    |
            |    +------+          |   +----+
            |    |      |          |    bbox2
            +----+------+----------+
                 |      |
                 +------+ 
                  bbox1
```
The internal method _mosaic_transform use following code to deal with it.
```python
            mosaic_bboxes[:, 0::2] = np.clip(mosaic_bboxes[:, 0::2], 0,
                                             2 * self.img_scale[1])
            mosaic_bboxes[:, 1::2] = np.clip(mosaic_bboxes[:, 1::2], 0,
                                             2 * self.img_scale[0])
```
After that, the bbox1 are cropped correctly, but the bbox2 are clipped into a line as shown in figure 2.
```python
             figure 2. before clip 
            +----------------------+
            |                      |
            |                      |    
            |                      + 
            |                      | bbox2  
            |    +------+          + 
            |    |      |          |
            +----+------+----------+
                  bbox1
```
That will lead to NaN loss during training.
## The modified code
The adding filtering rule aims to filter out the lines (such bbox2 in figure 2) to avoid NaN loss. So I have changed the default value to 0 but it will be executed when it is equal to 0.

Comment (zhaojinjian0000) [reply mmdet/datasets/pipelines/transforms.py]: @hhaAndroid  The above reply is the reason why the new code execute when the threshold is equal to 0.

Comment (hhaAndroid) [reply mmdet/datasets/pipelines/transforms.py]: Thank you for your detailed answer. Are you interested in adding the above content as docstr to the filtering rules?

Comment (zhaojinjian0000) [reply mmdet/datasets/pipelines/transforms.py]: Thank you for your advices. The above content does not need to be docstr because it describes how the bug occurs and the bug have been fixed.

Comment (hhaAndroid) [reply mmdet/datasets/pipelines/transforms.py]: If you are free to split the above content into 2 PRs, the first bug will mention a PR separately, and then we will merge it today.

Comment (hhaAndroid) [reply mmdet/datasets/pipelines/transforms.py]: Thank you very much

Comment (zhaojinjian0000) [reply mmdet/datasets/pipelines/transforms.py]: Thank you very much, I have modified the file and remove the fix of first bug in this PR at commit "fix this bug in another PR".

**Comment by hhaAndroid** [comment]
@zhaojinjian0000 Please fix the lint and then we will merge.

**Comment by zhaojinjian0000** [comment]
> @zhaojinjian0000 Please fix the lint and then we will merge.
@hhaAndroid , Thank you, and I have fixed the lint in "fix the lint" and "fix the spell".

**Comment by hhaAndroid** [comment]
> > @zhaojinjian0000 Please fix the lint and then we will merge.
> > @hhaAndroid , Thank you, and I have fixed the lint in "fix the lint" and "fix the spell".

ok

**Comment by ZwwWayne** [comment]
It occurs to me that we have a  FilterAnnotation pipeline, can we directly use it?

**Review by hhaAndroid — APPROVED** [review]

**Comment by hhaAndroid** [comment]
> It occurs to me that we have a FilterAnnotation pipeline, can we directly use it?

This pr can be merged first. Considering that there are other transformations that have this problem, it can be solved in a unified way later.

**Comment by OpenMMLab-Coodinator** [comment]
Hi @zhaojinjian0000 ！First of all, we want to express our gratitude for your significant PR in this project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/raweFPmdzG

If you are Chinese or have WeChat，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤
